### PR TITLE
Added SourceLocation to inherited chunks \ tag helpers

### LINKED ISSUE #2321: Give accurate location of syntax error in _GlobalImport.cshtml
Repro steps:
(1) Create a Starter Web App
(2) Turn off precompilation by deleting `Compiler` folder
(3) Create syntax error in `Views/_GlobalImport.cshtml` by `@using` a wrong namespace
(4) F5

The error message shows the error is in `Index.cshtml`, instead of `_GlobalImport.cshtml`:
capture

### PR DESCRIPTION
**Pull request by pranavkm**
Updated CompilationResult to support compilation failures from multiple
files.

Fixes #2321

### CODE REVIEW

**Comment by pranavkm** [comment]
Ties into Diagnostics changes - https://github.com/aspnet/Diagnostics/pull/123 and Razor changes - https://github.com/aspnet/Razor/pull/357. The error page from Helios already supports multiple error files.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs [hunk 1]
```diff
@@ -131,6 +125,56 @@ public CompilationResult Compile([NotNull] RelativeFileInfo fileInfo, [NotNull]
...
+            var diagnosticGroups = diagnostics
+                .Where(IsError)
+                .GroupBy(diagnostic => GetFilePath(relativePath, diagnostic), StringComparer.Ordinal);
+
+            var failures = new List<ICompilationFailure>();
+            foreach (var group in diagnosticGroups)
+            {
+                var sourceFilePath = group.Key;
+                string sourceFileContent;
+                if (string.Equals(assemblyName, sourceFilePath, StringComparison.Ordinal))
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: This allows us to show errors if they happen to be produced in the generated code (for instance a custom base type that does have a `RenderAsync` method). Mapping it to the cshtml file is not very helpful in diagnosing the issue.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs [hunk 2]
```diff
@@ -30,39 +36,61 @@ public CompilationResult Compile([NotNull] RelativeFileInfo file)
...
+        // Internal for unit testing
+        internal CompilationResult GetCompilationFailedResult(RelativeFileInfo file, IEnumerable<RazorError> errors)
         {
-            try
+            var messageGroups = errors
+                .GroupBy(razorError =>
+                // If a SourceLocation does not specify a file path, assume it is produced
+                // from parsing the current file.
+                razorError.Location.FilePath ?? file.RelativePath,
+                StringComparer.OrdinalIgnoreCase);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs]: Needs to be `Ordinal`

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkInheritanceUtility.cs [hunk 3]
```diff
@@ -126,22 +126,33 @@ public virtual IReadOnlyList<CodeTree> GetInheritedCodeTrees([NotNull] string pa
...
                     var codeGenerator = language.CreateCodeGenerator(className,
                                                                      engine.Host.DefaultNamespace,
-                                                                     viewStartPath,
+                                                                     globalImportPath,
                                                                      engine.Host);
                     codeGenerator.Visit(parseResults);
 
-                    return codeGenerator.Context.CodeTreeBuilder.CodeTree;
+                    var codeTree = codeGenerator.Context.CodeTreeBuilder.CodeTree;
+                    // Rewrite the location of inherited chunks so they point to the global import file.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkInheritanceUtility.cs]: nit: blank line before mid-block comments.  in this case, might be best to move this comment above previous line.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationFailedException.cs [hunk 4]
```diff
@@ -17,24 +17,24 @@ public class CompilationFailedException : Exception, ICompilationException
...
         {
-            CompilationFailures = new[] { compilationFailure };
+            CompilationFailures = compilationFailures;
         }
 
         /// <inheritdoc />
         public IEnumerable<ICompilationFailure> CompilationFailures { get; }
 
-        private static string FormatMessage(ICompilationFailure compilationFailure)
+        private static string FormatMessage(IEnumerable<ICompilationFailure> compilationFailure)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationFailedException.cs]: `compilationFailures` (add an 's')

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationResult.cs [hunk 5]
```diff
@@ -31,10 +32,10 @@ protected CompilationResult()
         public string CompiledContent { get; protected set; }
 
         /// <summary>
-        /// Gets the <see cref="ICompilationFailure"/> produced from parsing or compiling the Razor file.
+        /// Gets the <see cref="ICompilationFailure"/>s produced from parsing or compiling the Razor file.
         /// </summary>
         /// <remarks>This property is <c>null</c> when compilation succeeded.</remarks>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationResult.cs]: what does it mean if the collection is empty?

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationResult.cs]: It's still considered a compilation failure.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationResult.cs]: :ok: this comment should mention that case

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: that's a very general statement and doesn't seem correct.  if it were correct, Razor would not generate `#line` pragmas at all.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: or, are you commenting only on the case where the `#line` mapping is currently `hidden`?

File: src/Microsoft.AspNet.Mvc.Razor/Resources.resx [hunk 6]
```diff
@@ -189,4 +189,7 @@
   <data name="RazorFileInfoCollection_ResourceCouldNotBeFound" xml:space="preserve">
     <value>The resource '{0}' specified by '{1}' could not be found.</value>
   </data>
+  <data name="GeneratedCodeFileName" xml:space="preserve">
+    <value>Generated Code</value>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Resources.resx]: :smiley_cat:

File: test/Microsoft.AspNet.Mvc.Razor.Host.Test/Directives/ChunkInheritanceUtilityTest.cs [hunk 7]
```diff
@@ -39,28 +39,38 @@ @inherits MyBaseType
...
             // Assert
             Assert.Equal(2, codeTrees.Count);
-            var viewStartChunks = codeTrees[0].Chunks;
-            Assert.Equal(3, viewStartChunks.Count);
+            var globalImportChunks = codeTrees[0].Chunks;
+            Assert.Equal(3, globalImportChunks.Count);
 
-            Assert.IsType<LiteralChunk>(viewStartChunks[0]);
-            var usingChunk = Assert.IsType<UsingChunk>(viewStartChunks[1]);
+            var globalImportPath = @"Views\home\_GlobalImport.cshtml";
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Host.Test/Directives/ChunkInheritanceUtilityTest.cs]: should be in `// Arrange` or (worse) at least closer to its first use

**Comment by dougbu** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs [hunk 8]
```diff
@@ -16,12 +19,15 @@ public class RazorCompilationService : IRazorCompilationService
     {
         private readonly ICompilationService _compilationService;
         private readonly IMvcRazorHost _razorHost;
+        private readonly IFileProvider _fileProvider;
 
         public RazorCompilationService(ICompilationService compilationService,
-                                       IMvcRazorHost razorHost)
+                                       IMvcRazorHost razorHost,
+                                       IOptions<RazorViewEngineOptions> viewEngineOptions)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs]: curious: Why not just do `IFileProvider`?

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs]: The file system that we use for views specifically hangs off of `RazorViewEngineOptions`. It allows doing things like using resources as view sources - https://github.com/aspnet/Mvc/blob/dev/test/WebSites/RazorEmbeddedViewsWebSite/Startup.cs#L20-L23

**Comment by NTaylorMullen** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: Yes, this is for generated code that we produce but don't map back to the cshtml - this includes things like default inherited directives (`IHtmlHelper<TModel>`) and any code that relies on a property \ method being present on the base type `DefineSection`, `RenderAsync` etc). 

* I have a tracking bug for the first case - https://github.com/aspnet/Mvc/issues/2373.
* I wasn't able to find the bug the last time I looked - https://github.com/aspnet/Mvc/issues/1968. Mapping to the generated code should give you more more useful information for errors of this sort.

**Pull request closed without merging**